addons/html_editor/static/src/others/dynamic_placeholder/dynamic_placeholder.js

```diff
@@ -0,0 +1,113 @@
+    static name = "dynamic_placeholder";
+    static dependencies = ["overlay", "selection", "history", "dom", "qweb"];
+    /** @type { (p: DynamicPlaceholderPlugin) => Record<string, any> } */
+    static resources = (p) => ({
+        powerboxCategory: { id: "marketing_tools", name: _t("Marketing Tools"), sequence: 60 },
+        powerboxCommands: {
+            name: _t("Dynamic Placeholder"),
+            description: _t("Insert personalized content"),
+            category: "marketing_tools",
+            fontawesome: "fa-magic",
+            action(dispatch) {
+                dispatch("OPEN_DYNAMIC_PLACEHOLDER");
+            },
+        },
+    });
+    setup() {
+        this.defaultResModel = this.config.dynamicPlaceholderResModel;
+
+        /** @type {import("@html_editor/core/overlay_plugin").Overlay} */
+        this.overlay = this.shared.createOverlay(EditorDynamicPlaceholder, {
+            position: "bottom-start",
+        });
+        this.overlayProps = {
+            close: this.onClose.bind(this),
+            validate: this.onValidate.bind(this),
+        };
+
+        this.addDomListener(this.document, "click", () => {
+            if (this.overlay.isOpen) {
+                this.overlay.close();
+            }
+        });
+
+        this.addDomListener(
+            this.document,
+            "beforeinput",
+            (e) => {
+                if (this.overlay.isOpen && e.inputType === "insertParagraph") {
+                    e.preventDefault();
+                    e.stopImmediatePropagation();
+                }
+            },
+            true
+        );
+    }
+
+    handleCommand(command, payload) {
+        switch (command) {
+            case "OPEN_DYNAMIC_PLACEHOLDER": {
+                this.open(payload?.resModel || this.defaultResModel);
+                break;
+            }
+        }
+    }
+
+    /**
+     * @param {string} resModel
+     */
+    open(resModel) {
+        this.preservedSelection = this.shared.preserveSelection();
+        this.overlay.open({
+            props: {
+                ...this.overlayProps,
+                resModel: resModel,
+            },
+        });
+    }
+
+    /**
+     * @param {string} chain
+     * @param {string} defaultValue
+     */
+    onValidate(chain, defaultValue) {
+        if (!chain) {
+            return;
+        }
+
+        let dynamicPlaceholder = "object." + chain;
+        dynamicPlaceholder +=
+            defaultValue && defaultValue !== "" ? ` or '''${defaultValue}'''` : "";
+        const t = document.createElement("T");
+        t.setAttribute("t-out", dynamicPlaceholder);
+
+        this.shared.domInsert(t);
+        this.editable.focus();
+        this.dispatch("NORMALIZE", { node: t.parentElement });
+        this.dispatch("ADD_STEP");
+    }
+    onClose() {
+        this.preservedSelection?.restore();
+        delete this.preservedSelection;
+
+        setTimeout(() => this.overlay.close(), 50);
```

Comment:
Why do you need to wait 50 ms ?

Reply:
We need to prevent the enter create a insert paragraph in the editor. if I don't wait a litle bit here. the overlay is closed before the input event reach the event listener at line:42, an so we cannot prevent the event to reach the editor and create an unwanted break line.

I'm open to better solution if we can remove the time out tho.

Reply:
why not catch the "enter" keydown event?

Reply:
still the same q ? (You have some code to stop it no ?